pkg/overlay/selection_test.go

```diff
@@ -216,3 +219,119 @@ func TestNodeSelection(t *testing.T) {
 		}
 	})
 }
+
+func TestDistinctIPs(t *testing.T) {
+	testplanet.Run(t, testplanet.Config{
+		SatelliteCount: 1, StorageNodeCount: 10, UplinkCount: 1,
+		Reconfigure: testplanet.Reconfigure{
+			Satellite: func(log *zap.Logger, index int, config *satellite.Config) {
+				config.Discovery.RefreshInterval = 60 * time.Second
+				config.Discovery.DiscoveryInterval = 60 * time.Second
+				config.Discovery.GraveyardInterval = 60 * time.Second
+			},
+		},
+	}, func(t *testing.T, ctx *testcontext.Context, planet *testplanet.Planet) {
+		var err error
+		satellite := planet.Satellites[0]
+		service := satellite.Overlay.Service
+		tests := []struct {
+			nodeCount      int
+			duplicateCount int
+			requestCount   int
+			preferences    overlay.NodeSelectionConfig
+			shouldFailWith *errs.Class
+		}{
+			{ // test only distinct IPs with half new nodes
+				duplicateCount: 7,
+				requestCount:   4,
+				preferences: overlay.NodeSelectionConfig{
+					AuditCount:        8,
+					NewNodePercentage: 0.5,
+					OnlineWindow:      time.Hour,
+					DistinctIP:        true,
+				},
+			},
+			{
+				duplicateCount: 9,
+				requestCount:   2,
+				preferences: overlay.NodeSelectionConfig{
+					AuditCount:        0,
+					NewNodePercentage: 0,
+					OnlineWindow:      time.Hour,
+					DistinctIP:        true,
+				},
+			},
+			{ // test not enough distinct IPs
+				duplicateCount: 7,
+				requestCount:   7,
+				preferences: overlay.NodeSelectionConfig{
+					AuditCount:        0,
+					NewNodePercentage: 0,
+					OnlineWindow:      time.Hour,
+					DistinctIP:        true,
+				},
+				shouldFailWith: &overlay.ErrNotEnoughNodes,
+			},
+			{ // test distinct flag false allows duplicates
+				duplicateCount: 10,
+				requestCount:   5,
+				preferences: overlay.NodeSelectionConfig{
+					AuditCount:        0,
+					NewNodePercentage: 0.5,
+					OnlineWindow:      time.Hour,
+					DistinctIP:        false,
+				},
+			},
+		}
+
+		// This sets a reputable audit count for a certain number of nodes.
+		for i, node := range planet.StorageNodes {
```

Comment:
You shouldn't need to have this logic here. This double loop was intended to allow us to test all sorts of edge cases with reputation selection. For duplicate IPs, you just want to test selecting new vs. selecting reputable, so you can `UpdateStats` on half the nodes, and call those ones "reputable" (`preferences.AuditCount = 1`), and don't update the other half and call those ones "new".

Then you can set duplicate/non duplicate IPs accordingly to ensure you are covering the cases you care about.